activation/handler_v2.go

```diff
@@ -175,3 +182,69 @@ func (h *HandlerV2) syntacticallyValidate(ctx context.Context, atx *wire.Activat
 
 	return nil
 }
+
+// registerHashes registers that the given peer should be asked for
+// the hashes of the poet proofs and ATXs.
+func (h *HandlerV2) registerHashes(peer p2p.Peer, poetRefs []types.Hash32, atxIDs []types.ATXID) {
+	hashes := make([]types.Hash32, 0, len(atxIDs)+1)
+	for _, id := range atxIDs {
+		hashes = append(hashes, id.Hash32())
+	}
+	for _, poetRef := range poetRefs {
+		hashes = append(hashes, types.Hash32(poetRef))
+	}
+
+	h.fetcher.RegisterPeerHashes(peer, hashes)
+}
+
+// fetchReferences makes sure that the referenced poet proof and ATXs are available.
+func (h *HandlerV2) fetchReferences(ctx context.Context, poetRefs []types.Hash32, atxIDs []types.ATXID) error {
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for _, poetRef := range poetRefs {
+		eg.Go(func() error {
+			if err := h.fetcher.GetPoetProof(ctx, poetRef); err != nil {
+				return fmt.Errorf("fetching poet proof (%s): %w", poetRef.ShortString(), err)
+			}
+			return nil
+		})
+	}
+
+	if len(atxIDs) != 0 {
+		eg.Go(func() error {
+			if err := h.fetcher.GetAtxs(ctx, atxIDs, system.WithoutLimiting()); err != nil {
+				return fmt.Errorf("missing atxs %x: %w", atxIDs, err)
+			}
+			return nil
+		})
+	}
+	return eg.Wait()
```

Comment:
I don't think we need to use a Waitgroup here. Fetching the ATX in parallel to the referenced poet proof just results in the routine fetching the ATX to request (possibly) the same poet proof again. Even if the underlying fetcher properly handles multiple requests to the same reference correctly, this just puts more pressure on the scheduler managing extra go routines that otherwise wouldn't have been created.

```suggestion
	for _, poetRef := range poetRefs {
		if err := h.fetcher.GetPoetProof(ctx, poetRef); err != nil {
			return fmt.Errorf("fetching poet proof (%s): %w", poetRef.ShortString(), err)
		}
	}

	if len(atxIDs) == 0 {
		return nil
	}

	if err := h.fetcher.GetAtxs(ctx, atxIDs, system.WithoutLimiting()); err != nil {
		return fmt.Errorf("missing atxs %x: %w", atxIDs, err)
	}
	return nil
```

Wdyt?

Reply:
I did it this way because it has a very high chance of fetching everything in a single go (in a single batched hash request). The version you propose requires 3 requests (2xpoets + 1xatxs).